libraries/aws/shadow/src/shadow.c

```diff
@@ -0,0 +1,618 @@
+                                         uint16_t stringLength,
+                                         const char * pSubString,
+                                         uint16_t subStringLength )
+{
+    ShadowStatus_t returnStatus = SHADOW_FAIL;
+
+    /* The string must be at least as long as the substring to contain it
+     * completely. */
+    if( stringLength >= subStringLength )
+    {
+        /* We are only checking up to subStringLength characters in the original
+         * string. The string may be longer and contain additional characters. */
+        if( strncmp( pString, pSubString, ( size_t  ) subStringLength ) == 0 )
+        {
+            returnStatus = SHADOW_SUCCESS;
+        }
+    }
+
+    return returnStatus;
+}
+/*-----------------------------------------------------------*/
+
+static ShadowStatus_t extractThingName( const char * pString,
+                                        uint16_t stringLength,
+                                        uint16_t * pThingNameLength )
+{
+    uint16_t index = 0U;
+    ShadowStatus_t returnStatus = SHADOW_THINGNAME_PARSE_FAILED;
+
+    for( ; index < stringLength; index++ )
+    {
+        if( pString[ index ] == ( char ) '/' )
+        {
+            break;
+        }
+    }
+
+    /* Zero length thing name is not valid,
+     * $"$aws/things/<ThingName>/
+     * $"$aws/things/<ThingName>"
+     * will extract the same thing name result.
+     * Only empty thing name string like:
+     * "$aws/things/" or "$aws/things" will fail.
+     */
+    if( index > 0U )
+    {
+        * pThingNameLength = index;
+        returnStatus = SHADOW_SUCCESS;
+    }
+
+    return returnStatus;
+}
+/*-----------------------------------------------------------*/
+
+static ShadowStatus_t extractShadowMessageType( const char * pString,
+                                                uint16_t stringLength,
+                                                ShadowMessageType_t * pMessageType )
+{
+    uint32_t index = 0U;
+    ShadowStatus_t returnStatus = SHADOW_FAIL;
+
+    /* Lookup table for Shadow message string. */
+    static const char * const pMessageStrings[ ShadowMessageTypeMaxNum ] =
+    {
+        SHADOW_OP_GET_ACCEPTED,
+        SHADOW_OP_GET_REJECTED,
+        SHADOW_OP_DELETE_ACCEPTED,
+        SHADOW_OP_DELETE_REJECTED,
+        SHADOW_OP_UPDATE_ACCEPTED,
+        SHADOW_OP_UPDATE_REJECTED,
+        SHADOW_OP_UPDATE_DOCUMENTS,
+        SHADOW_OP_UPDATE_DELTA
+    };
+
+    /* Lookup table for Shadow message string length. */
+    static const uint16_t pMessageStringsLength[ ShadowMessageTypeMaxNum ] =
+    {
+        SHADOW_OP_GET_ACCEPTED_LENGTH,
+        SHADOW_OP_GET_REJECTED_LENGTH,
+        SHADOW_OP_DELETE_ACCEPTED_LENGTH,
+        SHADOW_OP_DELETE_REJECTED_LENGTH,
+        SHADOW_OP_UPDATE_ACCEPTED_LENGTH,
+        SHADOW_OP_UPDATE_REJECTED_LENGTH,
+        SHADOW_OP_UPDATE_DOCUMENTS_LENGTH,
+        SHADOW_OP_UPDATE_DELTA_LENGTH
+    };
```

Comment:
I mentioned this before, but it should be possible to reuse `getShadowOperationString` and `getShadowOperationLength` instead of using the above two arrays. You'd probably have to declare an array of `ShadowTopicStringType_t`s that are also message types, and then have a mapping of `ShadowTopicStringType_t` to `ShadowMessageType_t`, so it's up to you if you want to implement it.

Reply:
For extractShadowMessageType, we are walking through the table. It's not enum sensitive, so I think it's not required to implement it.
I'll resolve this one. 

If anyone thinks that we should implement it that way, feel free to reopen it.

Thank you